In [1]:
from importlib import reload
import pandas as pd
import time
import TextPreprocessing as wpp
import CountryExtractor as ce
import SentimentAnalyser as sa
import TopicCategorizer as tc

Reading raw data

In [19]:
data = pd.read_csv('../data/new_data_vdss.tsv.xz', sep='\t', compression='xz')

In [20]:
# Datum bereinigen & redundante Spalten löschen

data['datum'] = data['pubtime'].str.extract(r'(\d{4}-\d{2}-\d{2})')
data['datum'] = pd.to_datetime(data['datum'])
del data['pubtime']
del data['medium_code']
del data['regional']
del data['doctype']
del data['language']
del data['char_count']
del data['dateline']
del data['subhead']
del data['content_id']

Text preprocessing for content

In [21]:
# Headlines preprocessing with class word_preprocessing

preprocessor = wpp.headlinePreprocessing()

start_time = time.time()
data['content_processed'] = data['content'].apply(lambda x: preprocessor.preprocess_text(x) if isinstance(x, str) else [])
end_time = time.time()

elapsed_time = end_time - start_time

del data['content']

print("Elapsed Time: {:.2f} seconds".format(elapsed_time))

KeyboardInterrupt: 

In [22]:
# preparing blick.ch rubrics
data.loc[data['medium_name'] == 'blick.ch', 'rubric'] = data.loc[data['medium_name'] == 'blick.ch', 'rubric'].str.split(' - ').str[0]
filtered_data = data[data['medium_name'] == 'blick.ch']['rubric']
data

id     medium_name              rubric   
0       47612274        blick.ch          Wirtschaft  \
1       46153316          srf.ch  Sport - Mehr Sport   
2       45099545    swissinfo.ch          Gesundheit   
3       47381556  NZZ am Sonntag  Mensch und Medizin   
4       47256809          srf.ch         Radio SRF 1   
...          ...             ...                 ...   
153195  48273157  Berner Zeitung              Region   
153196  45361730      20 minuten   St. Gallen/Region   
153197  45122731        blick.ch             Politik   
153198  45429217  NZZ am Sonntag             Schweiz   
153199  46091106        blick.ch              People   

              doctype_description   
0                   Online medium  \
1                   Online medium   
2                   Online medium   
3       Regional weekly newspaper   
4                   Online medium   
...                           ...   
153195   Regional daily newspaper   
153196   Regional daily newspaper   
153197              Online medium   
153198  Regional weekly newspaper   
153199              Online medium   

                                                     head   
0            Schweden beschliesst starke Leitzinserhöhung  \
1       Handball-Nati: Schweizerinnen unterliegen Pole...   
2       "Wir befürchten, dass die Tabakinitiative die ...   
3                                                    News   
4                      Wenn Wassersportler fliegen lernen   
...                                                   ...   
153195     Warum Opfer von Online-Hetze oft machtlos sind   
153196                         Bürocontainer in Vollbrand   
153197           Bundespräsident Cassis zu Besuch in Wien   
153198                                   Classe politique   
153199  Mats Hummels feiert Ostern mit Cathy und Sohn ...   

                                                  content      datum  
0       <tx><ld><p>Die schwedische Notenbank stemmt si... 2022-09-20  
1       <tx><ld><p>Die Schweizer Handball-Nati der Fra... 2022-04-24  
2       <tx><ld><p>Wirtschaftskreise wehren sich gegen... 2022-01-11  
3       <tx><zt>Medizincannabis und Nikotin gehen Hand... 2022-08-28  
4       <tx><ld><p>Man sieht sie bald auf jedem Schwei... 2022-08-15  
...                                                   ...        ...  
153195  <tx><ld><p>Am Internet-Pranger Ein linksradika... 2022-11-24  
153196  <tx><p>ARBON. Aus noch unbekannten Gründen bra... 2022-02-07  
153197  <tx><ld><p>Die nächsten Schritte der Schweiz i... 2022-01-13  
153198  <tx><p>Ignazio Cassis, Globetrotter, twitterte... 2022-02-13  
153199  <tx><ld><p>Nach Mats Hummels Dating-Chaos in d... 2022-04-18  

[153200 rows x 7 columns]

Country analysis with running time estimation

In [23]:
extractor = ce.CountryExtractor()
%timeit extractor.get_country(data['content'][0])

68.1 µs ± 9.06 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [25]:
# extraction of countries
extractor = ce.CountryExtractor()
data['countries'] = data['content_processed'].apply(lambda x: extractor.get_country(x) if isinstance(x[0], str) else [])
data

id     medium_name              rubric   
0       47612274        blick.ch          Wirtschaft  \
1       46153316          srf.ch  Sport - Mehr Sport   
2       45099545    swissinfo.ch          Gesundheit   
3       47381556  NZZ am Sonntag  Mensch und Medizin   
4       47256809          srf.ch         Radio SRF 1   
...          ...             ...                 ...   
153195  48273157  Berner Zeitung              Region   
153196  45361730      20 minuten   St. Gallen/Region   
153197  45122731        blick.ch             Politik   
153198  45429217  NZZ am Sonntag             Schweiz   
153199  46091106        blick.ch              People   

              doctype_description   
0                   Online medium  \
1                   Online medium   
2                   Online medium   
3       Regional weekly newspaper   
4                   Online medium   
...                           ...   
153195   Regional daily newspaper   
153196   Regional daily newspaper   
153197              Online medium   
153198  Regional weekly newspaper   
153199              Online medium   

                                                     head      datum   
0            Schweden beschliesst starke Leitzinserhöhung 2022-09-20  \
1       Handball-Nati: Schweizerinnen unterliegen Pole... 2022-04-24   
2       "Wir befürchten, dass die Tabakinitiative die ... 2022-01-11   
3                                                    News 2022-08-28   
4                      Wenn Wassersportler fliegen lernen 2022-08-15   
...                                                   ...        ...   
153195     Warum Opfer von Online-Hetze oft machtlos sind 2022-11-24   
153196                         Bürocontainer in Vollbrand 2022-02-07   
153197           Bundespräsident Cassis zu Besuch in Wien 2022-01-13   
153198                                   Classe politique 2022-02-13   
153199  Mats Hummels feiert Ostern mit Cathy und Sohn ... 2022-04-18   

                                        content_processed   
0       [Die, schwedische, Notenbank, stemmt, kräftige...  \
1       [Die, Schweizer, Handball, Nati, Frauen, verli...   
2       [Wirtschaftskreise, wehren, starke, Einschränk...   
3       [Medizincannabis, Nikotin, gehen, Hand, Hand, ...   
4       [Man, sieht, bald, Schweizer, See, Wingfoiler,...   
...                                                   ...   
153195  [Am, Internet, Pranger, Ein, linksradikales, P...   
153196  [ARBON, Aus, unbekannten, Gründen, brannte, Na...   
153197  [Die, nächsten, Schritte, Schweiz, Europa, Pol...   
153198  [Ignazio, Cassis, Globetrotter, twitterte, Bil...   
153199  [Nach, Mats, Hummels, Dating, Chaos, vergangen...   

                                             countries  
0                                           [Schweden]  
1       [Ukraine, Polen, Russland, Slowenien, Schweiz]  
2                       [Frankreich, Schweiz, Italien]  
3                                                   []  
4                           [Schweiz, Australien, USA]  
...                                                ...  
153195                          [Schweiz, Irland, USA]  
153196                                              []  
153197   [Ukraine, Österreich, USA, Russland, Schweiz]  
153198                                         [Niger]  
153199                                    [Österreich]  

[153200 rows x 8 columns]

16.2 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

Sentiment analysis with running time estimation

In [31]:
reload(sa)
sentiment_analyser = sa.sentimentAnalyser()
%timeit sentiment_analyser.get_topic_sentiments_polarity(data['content_processed'][0])

65.9 ms ± 1.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


12.8 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [33]:
# Sentiment analysis

sentiment_analyser = sa.sentimentAnalyser()

start_time = time.time()
data['Sentiment'] = data['content_processed'].apply(lambda x: sentiment_analyser.get_topic_sentiments_polarity(x))
end_time = time.time()

elapsed_time = end_time - start_time

print("Elapsed Time: {:.2f} seconds".format(elapsed_time))

data.to_csv('../data/processed_data.tsv.xz', sep='\t', index=False, compression='xz')

Elapsed Time: 7932.10 seconds


Elapsed Time: 7932.10 seconds

Subjectivity analysis with running time estimation

In [73]:
reload(sa)
sentiment_analyser = sa.sentimentAnalyser()
%timeit sentiment_analyser.get_topic_subjectivity(processed_data['content_processed'][0])

16.4 ms ± 1.66 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [75]:
sentiment_analyser = sa.sentimentAnalyser()

start_time = time.time()

data['objectivity'] = data['content_processed'].apply(lambda x: sentiment_analyser.get_topic_subjectivity(x))

end_time = time.time()

elapsed_time = end_time - start_time

print("Elapsed Time: {:.2f} seconds".format(elapsed_time))

data.to_csv('../data/processed_data2.tsv.xz', sep='\t', index=False, compression='xz')

Elapsed Time: 9175.09 seconds


Elapsed Time: 9175.09 seconds

Text preprocessing 'rubric' and running time estimation

In [144]:
text_processing = wpp.TextPreprocessing()
%timeit data['rubric'][:1].apply(lambda x: text_processing.preprocess_text(x.lower()))

330 µs ± 16.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


330 µs ± 16.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [150]:
data['topic'] = data['rubric'].apply(lambda x: text_processing.preprocess_text(x.lower()) if isinstance(x, str) else [])

Import of processed data

In [38]:
processed_data = pd.read_csv('../data/processed_data2.tsv.xz', sep='\t', compression='xz')

In [40]:
# eval converts string to list
processed_data['content_processed'] = processed_data['content_processed'].apply(eval)

In [46]:
import TopicCategorizer as tc
reload(tc)
%timeit processed_data['content_processed'][:1].apply(lambda x: categorizer.categorize(x))

14 ms ± 899 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [ ]:
categorizer = tc.TopicCategorizer()

start_time = time.time()
processed_data['kategorie'] = processed_data['content_processed'].apply(lambda x: categorizer.categorize(x))
end_time = time.time()

elapsed_time = end_time - start_time

print("Elapsed Time: {:.2f} seconds".format(elapsed_time))

processed_data.to_csv('../data/processed_data3.tsv.xz', sep='\t', index=False, compression='xz')

In [45]:
processed_data

id     medium_name              rubric   
0       47612274        blick.ch          Wirtschaft  \
1       46153316          srf.ch  Sport - Mehr Sport   
2       45099545    swissinfo.ch          Gesundheit   
3       47381556  NZZ am Sonntag  Mensch und Medizin   
4       47256809          srf.ch         Radio SRF 1   
...          ...             ...                 ...   
153195  48273157  Berner Zeitung              Region   
153196  45361730      20 minuten   St. Gallen/Region   
153197  45122731        blick.ch             Politik   
153198  45429217  NZZ am Sonntag             Schweiz   
153199  46091106        blick.ch              People   

              doctype_description   
0                   Online medium  \
1                   Online medium   
2                   Online medium   
3       Regional weekly newspaper   
4                   Online medium   
...                           ...   
153195   Regional daily newspaper   
153196   Regional daily newspaper   
153197              Online medium   
153198  Regional weekly newspaper   
153199              Online medium   

                                                     head       datum   
0            Schweden beschliesst starke Leitzinserhöhung  2022-09-20  \
1       Handball-Nati: Schweizerinnen unterliegen Pole...  2022-04-24   
2       "Wir befürchten, dass die Tabakinitiative die ...  2022-01-11   
3                                                    News  2022-08-28   
4                      Wenn Wassersportler fliegen lernen  2022-08-15   
...                                                   ...         ...   
153195     Warum Opfer von Online-Hetze oft machtlos sind  2022-11-24   
153196                         Bürocontainer in Vollbrand  2022-02-07   
153197           Bundespräsident Cassis zu Besuch in Wien  2022-01-13   
153198                                   Classe politique  2022-02-13   
153199  Mats Hummels feiert Ostern mit Cathy und Sohn ...  2022-04-18   

                                        content_processed   
0       [Die, schwedische, Notenbank, stemmt, kräftige...  \
1       [Die, Schweizer, Handball, Nati, Frauen, verli...   
2       [Wirtschaftskreise, wehren, starke, Einschränk...   
3       [Medizincannabis, Nikotin, gehen, Hand, Hand, ...   
4       [Man, sieht, bald, Schweizer, See, Wingfoiler,...   
...                                                   ...   
153195  [Am, Internet, Pranger, Ein, linksradikales, P...   
153196  [ARBON, Aus, unbekannten, Gründen, brannte, Na...   
153197  [Die, nächsten, Schritte, Schweiz, Europa, Pol...   
153198  [Ignazio, Cassis, Globetrotter, twitterte, Bil...   
153199  [Nach, Mats, Hummels, Dating, Chaos, vergangen...   

                                                countries  Sentiment   
0                                            ['Schweden']   0.700000  \
1       ['Ukraine', 'Polen', 'Russland', 'Slowenien', ...  -0.140000   
2                    ['Frankreich', 'Schweiz', 'Italien']   0.150000   
3                                                      []   0.000000   
4                        ['Schweiz', 'Australien', 'USA']   0.194118   
...                                                   ...        ...   
153195                       ['Schweiz', 'Irland', 'USA']  -0.432558   
153196                                                 []  -0.850000   
153197  ['Ukraine', 'Österreich', 'USA', 'Russland', '...   0.350000   
153198                                          ['Niger']   0.000000   
153199                                     ['Österreich']   0.540000   

        objectivity               kategorie  
0          0.000000              Wirtschaft  
1          0.000000                   Sport  
2          0.075000                 Politik  
3          0.000000  Wissenschaft & Technik  
4          0.088235                   Sport  
...             ...                     ...  
153195     0.134884                     NaN  
153196     0.000000                     NaN  
153197 